client/service/src/client/call_executor.rs

```diff
@@ -199,6 +213,8 @@ where
 			None => {
 				let state_runtime_code = sp_state_machine::backend::BackendRuntimeCode::new(&state);
 				let runtime_code = state_runtime_code.runtime_code()?;
+				let runtime_code = self.wasm_overwrite.try_replace(runtime_code, &state)?;
```

Comment:
I think this feels a bit clunky and I wonder if we can come up with a smoother API. If `wasm_overwrite` could take only the `spec_version` maybe we could get away with `try_replace(spec_v, &state)`? It also feels weird to me that `try_replace` returns the input `code` if no override was found; can't we move that check elsewhere, perhaps into `runtime_code()`?
Maybe if you expanded a bit on the motivation it'll be clearer to me why it needs to be like this.

Having something like this would be nicer imo, (pseudo-code):
```
let runtime_code = {
  let spec_ver = code.spec_version();
  self.runtime_override
    .and_then(|rt_override| rt_override.getspec_ver, &state) )
    .unwrap_or_else(|| state_runtime_code.runtime_code()? )
}; 
```

Reply:
Changed this so that now WasmOverwrite just has a `get` method that returns `None` if it doesn't find an overwrite, where LocalCallExecutor gets the version for a blob, so it looks something like

```rust
let runtime_code = {
    let runtime_code = state_runtime_code.runtime_code();
    self.check_overwrites(runtime_code).unwrap_or(runtime_code)?
};
```

I think it's a bit odd that it's accepting a runtime code and then unwrapping into another runtime code, but it encapsulates calling for the version into it's own function on LocalCallExecutor now, where the actual `get` function only needs `spec_version` and `heap_pages`. We could go further and put the `check_overwrite` in `WasmOverwrite` too

`check_overwrite` accepts RuntimeCode again, for the convenience of having `code.heap_pages`. We could achieve a more similar API to the comment by using the `runtime_version` method on LocalCallExecutor and then getting heap_pages before calling `get` and would probably look more like

```rust
let runtime_code = {
    let code = state_runtime_code.runtime_code()?;
    let spec = self.runtime_version(id)?.spec_version;
    self.wasm_overwrite
        .as_ref()
        .map(|o| o.get(&spec, code.heap_pages))
        .flatten()
        .unwrap_or(code)
};
```

The reason to avoid calling `self.runtime_version` on LocalCallExecutor now in `check_overwrites` is to avoid requiring another type parameter (`Block: BlockT`) on all LocalCallExecutor types. However it is way more clear on what's happening